casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    val wildcards = List(F_, C_)
+    val knownMatches =
+      List(S0, R0).flatMap(s => (F0 :: C0 :: wildcards).map(r => s   -> r)) ++
+        List(S1, R1).flatMap(s => (F1 :: C1 :: wildcards).map(r => s -> r))
+
+    knownMatches.contains(a -> b)
+  }
+
+  def allMatch(left: Seq[Rho], right: Seq[Rho]) =
+    (for {
+      l <- left
+      r <- right
+    } yield matches(l, r)).fold(true)(_ && _)
+
+  def someMatch(left: Seq[Rho], right: Seq[Rho]) =
+    (for {
+      l <- left
+      r <- right
+    } yield matches(l, r)).fold(false)(_ || _)
+
+  def ifCommEvent(rhos: Seq[Rho])(f: (Rho, Rho) => Assertion) =
+    if (rhos.size == 2) {
+      val rho1 :: rho2 :: scala.collection.immutable.Nil = rhos.toList
+      f(rho1, rho2)
+    }
+
+  /***
+   Two incoming sends/receives, at most one had a matching dual in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They couldn't be competing for the same linear receive/send (at most one had a match).
+   Notice this includes "two unsatisfied" and "must be looking for different data" cases.
+    */
+  object SamePolarityMerge extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   Two incoming sends/receives each matched a receive/send that was in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They could've matched the same linear event.
+   Mergeable if different events, or at least one matched event is non-linear.
+   This is the case where both incoming events could have matched what was in the other TS.
+    */
+  object CouldMatchSameConflicts extends ConflictingCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(allMatch(left ++ right, base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   Two incoming sends/receives each matched a receive/send that was in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They could've matched the same linear event.
+   Mergeable if different events, or at least one matched event is non-linear.
+   This is the case where the incoming events match differently
+    */
+  object CouldMatchSameMerges extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base) || onePersistent(base))
```

Comment:
This should be `allMatch(left, base) && allMatch(right, base`), + an assertion that they've either matched different events, or the matched-by-both event is non-linear.

I can't make a connection between the assertion in code ane the mergability rules/cases we're testing.

if we had signular `left` and `right`, this could be:

```
val m1 = findMatch(left, base) //throws asseriton error if not found
val m2 = findMatch(right, base) // we could also return Option-s and do the needed assertions
assert(m1 != m2 || m1.cardinality == NonLinear || m2.cardinality == NonLinear)
```

Notice the `Conflicts` case is exactly the same, just with the last assertion negated.
(`m1 == m2 && m1.cardinality == Linear && m2.cardinality == Linear`)